server/sms/sms_gateway_test.go

```diff
@@ -231,3 +233,104 @@ func TestReadLastID(t *testing.T) {
 
 	a.Equal(uint64(10), gw.LastIDSent)
 }
+
+func tempFilename() string {
+	file, err := ioutil.TempFile("/tmp", "guble_sms_retry")
+	if err != nil {
+		panic(err)
+	}
+	file.Close()
+	os.Remove(file.Name())
+	return file.Name()
+}
+
+func Test_RetryLoop(t *testing.T) {
+	ctrl, finish := testutil.NewMockCtrl(t)
+	defer finish()
+
+	defer testutil.EnableDebugForMethod()()
+	a := assert.New(t)
+
+	//create a mockSms Sender to simulate Nexmo error
+	mockSmsSender := NewMockSender(ctrl)
+	f := tempFilename()
+	defer os.Remove(f)
+
+	//create a KVStore with sqlite3
+	kvStore := kvstore.NewSqliteKVStore(f, false)
+	err := kvStore.Open()
+	a.NoError(err)
+	a.NotNil(kvStore)
+
+	//create a new mock router
+	routerMock := NewMockRouter(testutil.MockCtrl)
+	routerMock.EXPECT().KVStore().AnyTimes().Return(kvStore, nil)
+
+	//create a new mock msg store
+	mockMessageStore := NewMockMessageStore(ctrl)
+	routerMock.EXPECT().MessageStore().AnyTimes().Return(mockMessageStore, nil)
+	mockMessageStore.EXPECT().MaxMessageID(gomock.Any()).Return(uint64(2), nil)
+
+	//setup a new sms gateway
+	worker := 8
+	topic := SMSDefaultTopic
+	enableMetrics := false
+	gateway, err := New(routerMock, mockSmsSender, Config{Workers: &worker, Name: SMSDefaultTopic, Schema: SMSSchema, SMSTopic: &topic, IntervalMetrics: &enableMetrics})
+	a.NoError(err)
+
+	//create a new route on which the gateway will subscribe on /sms
+	route := router.NewRoute(router.RouteConfig{
+		Path:         protocol.Path(*gateway.config.SMSTopic),
+		ChannelSize:  5000,
+		FetchRequest: gateway.fetchRequest(),
+	})
+	//expect 2 subscribe.One for Start.One for Restart
+	routerMock.EXPECT().Subscribe(gomock.Any()).Times(2).Return(route, nil)
+
+	err = gateway.Start()
+	a.NoError(err)
+
+	//create a new sms message
+	sms := NexmoSms{
+		To:   "toNumber",
+		From: "FromNUmber",
+		Text: "body",
+	}
+	d, err := json.Marshal(&sms)
+	a.NoError(err)
+
+	msg := protocol.Message{
+		Path:          protocol.Path(topic),
+		UserID:        "samsa",
+		ApplicationID: "sms",
+		ID:            uint64(4),
+		Body:          d,
+	}
+
+	// when sms is sent simulate an error.No Sms will be delivered with success for 4 times.
+	mockSmsSender.EXPECT().Send(gomock.Eq(&msg)).Times(4).Return(ErrIncompleteSMSSent)
```

Comment:
https://logging-prod.rewe-digital.com/kibana4/#/discover?_g=(refreshInterval:(display:Off,pause:!f,section:0,value:0),time:(from:now-7d,mode:quick,to:now))&_a=(columns:!(wwIdent,sms_message,sms_scenario,sms_rule,sms_rule_case,module,msg,service),index:%5Btechnical-%5DYYYY.MM.DD,interval:auto,query:(query_string:(analyze_wildcard:!t,query:%27%22Nexmo%20SMS:%20enabled%22%27)),sort:!(%27@timestamp%27,desc))